include/open62541/util.h

```diff
@@ -47,27 +54,84 @@ typedef enum {
 
 /* Returns a pointer to the value or NULL if the key is not found.*/
 UA_EXPORT const UA_Variant *
-UA_KeyValueMap_get(const UA_KeyValuePair *map, size_t mapSize,
+UA_KeyValueMap_get(const UA_KeyValueMap *map,
                    const UA_QualifiedName key);
 
+/* Wether or not the given key-value-map is empty.
+ * Returns true if map is NULL or map->mapSize is equal to zero,
+ * otherwise false.
+ * */
+UA_EXPORT UA_Boolean
+UA_KeyValueMap_isEmpty(const UA_KeyValueMap *map);
+
 /* Returns NULL if the value for the key is not defined or not of the right
  * datatype and scalar/array */
 UA_EXPORT const void *
-UA_KeyValueMap_getScalar(const UA_KeyValuePair *map, size_t mapSize,
+UA_KeyValueMap_getScalar(const UA_KeyValueMap *map,
                          const UA_QualifiedName key,
                          const UA_DataType *type);
 
 /* Remove a single entry. To delete the entire map, use UA_Array_delete. */
 UA_EXPORT UA_StatusCode
-UA_KeyValueMap_delete(UA_KeyValuePair **map, size_t *mapSize,
+UA_KeyValueMap_remove(UA_KeyValueMap *map,
                       const UA_QualifiedName key);
 
+UA_EXPORT UA_KeyValueMap *
+UA_KeyValueMap_new(size_t size_hint);
+
+UA_EXPORT void
+UA_KeyValueMap_delete(UA_KeyValueMap *map);
+
+/**
+ * Create a deep copy of the given KeyValueMap. Function will allocate memory for
+ * the newly create KeyValueMap which needs to be freed by the user of the function
+ * accordingly.
+ * @param map The KeyValueMap to be copeid
+ * @param mapSize The size of the KeyValueMap
+ * @return A pointer to the copied KeyValueMap, NULL if memory allocation failed.
+ */
+UA_EXPORT UA_KeyValueMap*
+UA_KeyValueMap_copy(UA_KeyValueMap *map);
+
+/**
+ * Check KeyValueMap for containment
+ * ---------------------------------
+ * Check if a given KeyValueMap contains an entry mapped by the provided key.
+ * @param map The KeyValueMap to check against
+ * @param mapSize The size of the KeyValueMap
+ * @param key The key to check for existence in the KeyValueMap
+ * @return true if an entry with the given key exists, false otherwise
+ */
+UA_EXPORT UA_Boolean
+UA_KeyValueMap_contains(UA_KeyValueMap *map, UA_QualifiedName key);
+
+/**
+ * Merge two KeyValueMaps
+ * ----------------------
+ * The two key value maps will be merged where values from the left operand
+ * will be overwritten by values of the right operand when the keys are identical.
+ *
+ * @param dst The KeyValueMap to store the resulting merge in. The size can either be an upper bound (rhsCout + lhsCount)
+ *            or can be calculated beforehand by using UA_KeyValueMap_countMergedMembers
+ * @param lhs The left hand side operand of the KeyValueMap merge
+ * @param lhsCount The count of entries in the lhs KeyValueMap
+ * @param rhs The right hand side operand of the KeyValueMap merge
+ * @param rhsCount The count of entries in the rhs KeyValueMap
+ */
+UA_EXPORT void
+UA_KeyValueMap_merge(UA_KeyValueMap *lhs, UA_KeyValueMap *rhs);
```

Comment:
Should return a statuscode.
Make sure that - if the statuscode is bad - the input structs come back untouched.
The rhs can be const.